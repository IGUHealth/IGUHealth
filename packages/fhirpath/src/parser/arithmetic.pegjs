// Grammar rules
// [FHIRPath](http://hl7.org/fhirpath/N1) Normative Release

//prog: line (line)*;
//line: ID ( '(' expr ')') ':' expr '\r'? '\n';

expression
        = term expression_2
        / ('+' / '-') expression //polarityExpression
        
        /// (IDENTIFIER)? '=>' expression                           //lambdaExpression

expression_2
        = '.' invocation expression_2                            //invocationExpression
        / '[' expression ']' expression_2                        //indexerExpression
        / ('*' / '/' / 'div' / 'mod') expression expression_2    //multiplicativeExpression
        / ('+' / '-' / '&') expression  expression_2             //additiveExpression
       // / expression ('is' / 'as') //typeSpecifier             //typeExpression
        /  '/' expression  expression_2                          //unionExpression
        /  ('<=' / '<' / '>' / '>=') expression expression_2     //inequalityExpression
        /  ('=' / '~' / '!=' / '!~') expression expression_2     //equalityExpression
        /  ('in' / 'contains') expression expression_2           //membershipExpression
        /  'and' expression expression_2                         //andExpression
        /  ('or' / 'xor') expression expression_2                //orExpression
        /  'implies' expression expression_2                     //impliesExpression;

term
        = invocation                                            //invocationTerm
        / literal                                               //literalTerm
        / externalConstant                                      //externalConstantTerm
        / '(' expression ')'                                    //parenthesizedTerm
        ;


literal
        = '{' '}'                                               //nullLiteral
        / ('true' / 'false')                                    //booleanLiteral
        // / STRING                                                //stringLiteral
        // / NUMBER                                                //numberLiteral
        // / DATE                                                  //dateLiteral
        // / DATETIME                                              //dateTimeLiteral
        // / TIME                                                  //timeLiteral
        // / quantity                                              //quantityLiteral

externalConstant
        = '%' ( identifier / STRING )
        ;        

invocation                          // Terms that can be used after the function/member invocation '.'
        // = identifier                                            //memberInvocation
        // / function                                              //functionInvocation
        // / '$this'                                               //thisInvocation
        // / '$index'                                              //indexInvocation
        = '$total'                                              //totalInvocation        

identifier
        = IDENTIFIER
        /// DELIMITEDIDENTIFIER
        / 'as'
        / 'contains'
        / 'in'
        / 'is'
        ;        

ESC  = [^\'\"]

STRING
        = '\'' (ESC / .)* '\''
        ;

IDENTIFIER
        = ([A-Za-z] / '_')([A-Za-z0-9] / '_')*            // Added _ to support CQL (FHIR could constrain it out)
        ;