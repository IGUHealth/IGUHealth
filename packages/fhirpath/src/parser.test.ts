import { parse } from "./parser";

test("SNAPSHOT Tests", () => {
  // Operator testing
  expect(parse("$this.test + $this.test2")).toMatchSnapshot();
  expect(parse("$this.test + $this.test2 - 14")).toMatchSnapshot();
  expect(
    parse(
      "45 < $this.test.that * $this.test2 / 14 | 'test' in 'hello' and x != 12 and 45 implies 23"
    )
  ).toMatchSnapshot();

  // Deliniated identifiers testing
  expect(
    parse("test.`HELLO-WORLD IDENTIFIER`.deliniated.`strange-synt`")
  ).toMatchSnapshot();

  // expect(parse("45days - 12years")).toMatchSnapshot();
  // Parameter testing
  expect(
    parse(
      "$this.test.where(value != 'test' and value / 11 = 12, $index != $total + %testing)"
    )
  ).toMatchSnapshot();
  // Literal testing
  expect(parse("'testing'")).toMatchSnapshot();
  expect(
    parse(
      "true != false and @T04:00 = @T04:00 and @1980-01-01 != @1980-01-01T00:00:00Z = @1980-01-01T00:00:00-05:00"
    )
  ).toMatchSnapshot();
  // Number testing
  expect(parse("45 - 55.23 + 0.23")).toMatchSnapshot();

  expect(parse("this[45 + 15 * 3].hello")).toMatchSnapshot();
  expect(parse("$this.test + 2 * 4")).toMatchSnapshot();

  expect(parse("true.anyTrue()")).toMatchSnapshot();
});
