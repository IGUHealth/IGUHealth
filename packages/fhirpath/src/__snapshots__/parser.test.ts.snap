// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT Tests 1`] = `
{
  "left": {
    "next": [
      {
        "type": "Invocation",
        "value": {
          "type": "Identifier",
          "value": "test",
        },
      },
    ],
    "type": "Term",
    "value": {
      "type": "Invocation",
      "value": {
        "type": "This",
      },
    },
  },
  "operator": "+",
  "right": {
    "next": [
      {
        "type": "Invocation",
        "value": {
          "type": "Identifier",
          "value": "test2",
        },
      },
    ],
    "type": "Term",
    "value": {
      "type": "Invocation",
      "value": {
        "type": "This",
      },
    },
  },
  "type": "Operation",
}
`;

exports[`SNAPSHOT Tests 2`] = `
{
  "left": {
    "left": {
      "next": [
        {
          "type": "Invocation",
          "value": {
            "type": "Identifier",
            "value": "test",
          },
        },
      ],
      "type": "Term",
      "value": {
        "type": "Invocation",
        "value": {
          "type": "This",
        },
      },
    },
    "operator": "+",
    "right": {
      "next": [
        {
          "type": "Invocation",
          "value": {
            "type": "Identifier",
            "value": "test2",
          },
        },
      ],
      "type": "Term",
      "value": {
        "type": "Invocation",
        "value": {
          "type": "This",
        },
      },
    },
    "type": "Operation",
  },
  "operator": "-",
  "right": {
    "type": "Term",
    "value": {
      "type": "Literal",
      "value": 14,
    },
  },
  "type": "Operation",
}
`;

exports[`SNAPSHOT Tests 3`] = `
{
  "left": {
    "left": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 45,
      },
    },
    "operator": "<",
    "right": {
      "left": {
        "left": {
          "next": [
            {
              "type": "Invocation",
              "value": {
                "type": "Identifier",
                "value": "test",
              },
            },
            {
              "type": "Invocation",
              "value": {
                "type": "Identifier",
                "value": "that",
              },
            },
          ],
          "type": "Term",
          "value": {
            "type": "Invocation",
            "value": {
              "type": "This",
            },
          },
        },
        "operator": "*",
        "right": {
          "next": [
            {
              "type": "Invocation",
              "value": {
                "type": "Identifier",
                "value": "test2",
              },
            },
          ],
          "type": "Term",
          "value": {
            "type": "Invocation",
            "value": {
              "type": "This",
            },
          },
        },
        "type": "Operation",
      },
      "operator": "/",
      "right": {
        "left": {
          "type": "Term",
          "value": {
            "type": "Literal",
            "value": 14,
          },
        },
        "operator": "|",
        "right": {
          "left": {
            "type": "Term",
            "value": {
              "type": "Literal",
              "value": "test",
            },
          },
          "operator": "in",
          "right": {
            "left": {
              "type": "Term",
              "value": {
                "type": "Literal",
                "value": "hello",
              },
            },
            "operator": "and",
            "right": {
              "type": "Term",
              "value": {
                "type": "Invocation",
                "value": {
                  "type": "Identifier",
                  "value": "x",
                },
              },
            },
            "type": "Operation",
          },
          "type": "Operation",
        },
        "type": "Operation",
      },
      "type": "Operation",
    },
    "type": "Operation",
  },
  "operator": "!=",
  "right": {
    "left": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 12,
      },
    },
    "operator": "and",
    "right": {
      "left": {
        "type": "Term",
        "value": {
          "type": "Literal",
          "value": 45,
        },
      },
      "operator": "implies",
      "right": {
        "type": "Term",
        "value": {
          "type": "Literal",
          "value": 23,
        },
      },
      "type": "Operation",
    },
    "type": "Operation",
  },
  "type": "Operation",
}
`;

exports[`SNAPSHOT Tests 4`] = `
{
  "next": [
    {
      "type": "Invocation",
      "value": {
        "type": "Identifier",
        "value": "HELLO-WORLD IDENTIFIER",
      },
    },
    {
      "type": "Invocation",
      "value": {
        "type": "Identifier",
        "value": "deliniated",
      },
    },
    {
      "type": "Invocation",
      "value": {
        "type": "Identifier",
        "value": "strange-synt",
      },
    },
  ],
  "type": "Term",
  "value": {
    "type": "Invocation",
    "value": {
      "type": "Identifier",
      "value": "test",
    },
  },
}
`;

exports[`SNAPSHOT Tests 5`] = `
{
  "next": [
    {
      "type": "Invocation",
      "value": {
        "type": "Identifier",
        "value": "test",
      },
    },
    {
      "type": "Invocation",
      "value": {
        "next": [
          {
            "left": {
              "left": {
                "type": "Term",
                "value": {
                  "type": "Invocation",
                  "value": {
                    "type": "Identifier",
                    "value": "value",
                  },
                },
              },
              "operator": "!=",
              "right": {
                "left": {
                  "left": {
                    "type": "Term",
                    "value": {
                      "type": "Literal",
                      "value": "test",
                    },
                  },
                  "operator": "and",
                  "right": {
                    "type": "Term",
                    "value": {
                      "type": "Invocation",
                      "value": {
                        "type": "Identifier",
                        "value": "value",
                      },
                    },
                  },
                  "type": "Operation",
                },
                "operator": "/",
                "right": {
                  "type": "Term",
                  "value": {
                    "type": "Literal",
                    "value": 11,
                  },
                },
                "type": "Operation",
              },
              "type": "Operation",
            },
            "operator": "=",
            "right": {
              "type": "Term",
              "value": {
                "type": "Literal",
                "value": 12,
              },
            },
            "type": "Operation",
          },
          {
            "left": {
              "type": "Term",
              "value": {
                "type": "Invocation",
                "value": {
                  "type": "Index",
                },
              },
            },
            "operator": "!=",
            "right": {
              "left": {
                "type": "Term",
                "value": {
                  "type": "Invocation",
                  "value": {
                    "type": "Total",
                  },
                },
              },
              "operator": "+",
              "right": {
                "type": "Term",
                "value": {
                  "type": "Variable",
                  "value": {
                    "type": "Identifier",
                    "value": "testing",
                  },
                },
              },
              "type": "Operation",
            },
            "type": "Operation",
          },
        ],
        "type": "Function",
        "value": {
          "type": "Identifier",
          "value": "where",
        },
      },
    },
  ],
  "type": "Term",
  "value": {
    "type": "Invocation",
    "value": {
      "type": "This",
    },
  },
}
`;

exports[`SNAPSHOT Tests 6`] = `
{
  "type": "Term",
  "value": {
    "type": "Literal",
    "value": "testing",
  },
}
`;

exports[`SNAPSHOT Tests 7`] = `
{
  "left": {
    "left": {
      "left": {
        "left": {
          "type": "Term",
          "value": {
            "type": "Invocation",
            "value": {
              "type": "Identifier",
              "value": "true",
            },
          },
        },
        "operator": "!=",
        "right": {
          "left": {
            "type": "Term",
            "value": {
              "type": "Invocation",
              "value": {
                "type": "Identifier",
                "value": "false",
              },
            },
          },
          "operator": "and",
          "right": {
            "type": "Term",
            "value": {
              "type": "Literal",
              "value": "04:00",
            },
          },
          "type": "Operation",
        },
        "type": "Operation",
      },
      "operator": "=",
      "right": {
        "left": {
          "type": "Term",
          "value": {
            "type": "Literal",
            "value": "04:00",
          },
        },
        "operator": "and",
        "right": {
          "type": "Term",
          "value": {
            "type": "Literal",
            "value": "1980-01-01",
          },
        },
        "type": "Operation",
      },
      "type": "Operation",
    },
    "operator": "!=",
    "right": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": "1980-01-01T00:00:00Z",
      },
    },
    "type": "Operation",
  },
  "operator": "=",
  "right": {
    "type": "Term",
    "value": {
      "type": "Literal",
      "value": "1980-01-01T00:00:00-05:00",
    },
  },
  "type": "Operation",
}
`;

exports[`SNAPSHOT Tests 8`] = `
{
  "left": {
    "left": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 45,
      },
    },
    "operator": "-",
    "right": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 55.23,
      },
    },
    "type": "Operation",
  },
  "operator": "+",
  "right": {
    "type": "Term",
    "value": {
      "type": "Literal",
      "value": 0.23,
    },
  },
  "type": "Operation",
}
`;

exports[`SNAPSHOT Tests 9`] = `
{
  "next": [
    {
      "type": "Indexed",
      "value": {
        "left": {
          "type": "Term",
          "value": {
            "type": "Literal",
            "value": 45,
          },
        },
        "operator": "+",
        "right": {
          "left": {
            "type": "Term",
            "value": {
              "type": "Literal",
              "value": 15,
            },
          },
          "operator": "*",
          "right": {
            "type": "Term",
            "value": {
              "type": "Literal",
              "value": 3,
            },
          },
          "type": "Operation",
        },
        "type": "Operation",
      },
    },
    {
      "type": "Invocation",
      "value": {
        "type": "Identifier",
        "value": "hello",
      },
    },
  ],
  "type": "Term",
  "value": {
    "type": "Invocation",
    "value": {
      "type": "Identifier",
      "value": "this",
    },
  },
}
`;

exports[`SNAPSHOT Tests 10`] = `
{
  "left": {
    "next": [
      {
        "type": "Invocation",
        "value": {
          "type": "Identifier",
          "value": "test",
        },
      },
    ],
    "type": "Term",
    "value": {
      "type": "Invocation",
      "value": {
        "type": "This",
      },
    },
  },
  "operator": "+",
  "right": {
    "left": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 2,
      },
    },
    "operator": "*",
    "right": {
      "type": "Term",
      "value": {
        "type": "Literal",
        "value": 4,
      },
    },
    "type": "Operation",
  },
  "type": "Operation",
}
`;
